pkg/rbd/controllerserver.go

```diff
@@ -272,12 +262,8 @@ func (cs *ControllerServer) DeleteVolume(ctx context.Context, req *csi.DeleteVol
 		// If error is ErrImageNotFound then we failed to find the image, but found the imageOMap
 		// to lead us to the image, hence the imageOMap needs to be garbage collected, by calling
 		// unreserve for the same
-		volumeNameMutex.LockKey(rbdVol.RequestName)
-		defer func() {
-			if err := volumeNameMutex.UnlockKey(rbdVol.RequestName); err != nil {
-				klog.Warningf("failed to unlock mutex volume:%s %v", rbdVol.RequestName, err)
-			}
-		}()
+		idLk := volumeNameLocker.Lock(rbdVol.RequestName)
```

Comment:
can we lock the resource before fetching volume info from rados omap?
as a corner case if there is a volume  request with the same name created at the same we fetch the same info may clash, better to take a lock and fetch  omap

Reply:
(relevant to DeleteSnapshot as well, hence not reposting against the same)

We need `requestName` to lock, that is fetched from the RADOS OMaps, hence we cannot lock before fetching the information.

We hence need to analyze, is the post lookup locking of OMap safe.

1: Failing to fetch the UUID OMap during lookup
 - This means that the UUID is deleted, so we respond with an ack, not a problem

2: Fetching the incorrect UUID Omap
- We fetch the UUID OMap information, based on the UUID encoded on the `VolumeID`, the CO can only have this `VolumeID` if there was a prior successful `CreateVolume` call
- In the `CreateVolume` the UUID is generated to ensure there are no clashes with existing UUIDs on the pool, IOW we go beyond the usual "UUIDs are unique enough statistically" statements during its generation, to ensure it's uniqueness.
- So we will not fetch an incorrect UUID, and we further check if the UUID has the same `requestName` as the request name directory

3: Fetching an in flight UUID OMap
This case is slightly different from (2) and is a **problem**.
- Assume a `VolID` was just deleted by the CO, but the `DeleteVolume` response timed out at the CO, hence the CO would replay he same `VolID` for deletion
- Later a `CreateVolume` (for a different `RequestName`) used the just deleted UUID as it's UUID (this call can complete or be in-flight for the rest of the discussion)

Now, the next `DeleteVolume` for the older `VolID` would garbage collect the new UUID based RADOS entries and images.

The `DeleteVolume` could get into a situation where it finds the OMaps but the image still missing (as `CreateVolume` that is racing with this, and has not created the image yet) and only garbage collect the RADOS entries, this needs to be protected using an **UUID lock**, so that we lookup/create without conflicts.

BUT (and this is a big one), `DeleteVolume` will still garbage collect the new UUID, as once the Create is complete all entries align (`RequestName` OMAP and UUID OMap and image etc.). What this essentially means is that we cannot reuse a UUID as we cannot be sure `DeleteVolume` **caller** ever processed a success return from the plugin.

The ONLY resolution for the same would be to retain the RADOS UUID OMaps so that these are never reused (space consumption issues?), hence VolIDs are always unique? Or, never garbage collect the `RequestName` in the CSI directories OMap to ensure the the same request name gets the same UUID, slightly better but assumes request names are reused (which is not true in kubernetes as it uses a UUID for the request names anyway).

Also, it is interesting how other CSI vendors handle the same, as the problem is outside the scope of Ceph-CSI in reality, and should impact other implementations as well.

Reply:
Leaving my thought on the UUID collision problem I alluded to in the previous comment:
- We need to depend on [UUID collision](https://en.wikipedia.org/wiki/Universally_unique_identifier#Collisions) properties to satisfy this requirement, as otherwise we need to remember and not use any older UUID that was ever used
- We run a single instance of the provisioner plugin, hence we are at least not generating UUIDs across different systems for time stamp based or MAC address based collision probabilities to increase
- If we intend to solve this, it also has to be understood why others do not have this issue or avoid it (i.e never reuse an older `VolID`)

**IOW, I think this problem can be left unsolved.**

If there are alternative thoughts on resolving this do share.